lib/spack/spack/caches.py

```diff
@@ -51,24 +52,35 @@ def _fetch_cache():
 
 class MirrorCache(object):
     def __init__(self, root, skip_unstable_versions):
-        self.root = os.path.abspath(root)
+        self.root = spack.util.url.parse(root)
         self.skip_unstable_versions = skip_unstable_versions
 
     def store(self, fetcher, relative_dest):
         """Fetch and relocate the fetcher's target into our mirror cache."""
 
         # Note this will archive package sources even if they would not
         # normally be cached (e.g. the current tip of an hg/git branch)
-        dst = os.path.join(self.root, relative_dest)
+        dst = spack.util.url.local_file_path(
+            spack.util.url.join(self.root, relative_dest))
+
+        if not dst:
+            raise spack.error.SpackError(
+                'Cannot store sources into non-remote (file://) URLs')
```

Comment:
I left a general comment about this but we have to allow the fetcher to write here or the data is never pushed to remote mirrors on creation. Is there a reason you are forcing this to be a local file path? The fetcher contains the logic for how to archive the cached file, I think including any directories that need to exist first.